csharp/src/Ice/VectoredBuffer.cs

```diff
@@ -0,0 +1,549 @@
+            set
+            {
+                int index = value.Segment;
+                _currentSegmentOffset = 0;
+                for (int i = 0; i < index; i++)
+                {
+                    _currentSegmentOffset += Segments[i].Count;
+                }
+                _currentSegment = Segments[index];
+
+                if (value.Offset > _currentSegment.Count)
+                {
+                    throw new ArgumentException("The position offset is invalid for the current buffer", nameof(value));
+                }
+                _position = value;
+            }
+        }
+
+        /// <summary>Number of bytes from current position to the end.</summary>
+        public int Remaining => Size - (_position.Offset + _currentSegmentOffset);
+
+        /// <summary>Determines the current size of the stream, this correspond
+        /// to the number of bytes already writen to the stream.</summary>
+        /// <value>The current size.</value>
+        public int Size { get; private set; }
+
+        /// <summary>List of array segments own by the buffer, the last segment size reflects the size
+        /// of the underlying array until Prepare is call after that point it reflects the portion
+        /// of the last segment that is in use.</summary>
+        public IList<ArraySegment<byte>> Segments { get; }
+        // The segment curretnly used by write operations
+        private ArraySegment<byte> _currentSegment;
+        // Cached offset of the current segment
+        private int _currentSegmentOffset;
+        private const int MinSegmentSize = 256;
+        private BufferPosition _position;
+
+        /// <summary>
+        /// Create a new buffer if the initialData param is not null it will be used as the buffer
+        /// first segment, otherwise and empty segment will be created as the first segment.</summary>
+        /// <param name="initialData">Optional initial data for the buffer.</param>
+        public VectoredBuffer(byte[]? initialData = null)
+        {
+            Segments = new List<ArraySegment<byte>>();
+            _position = new BufferPosition(0, 0);
+            if (initialData == null)
+            {
+                Segments.Add(new ArraySegment<byte>(new byte[MinSegmentSize]));
+                Size = 0;
+                Capacity = MinSegmentSize;
+            }
+            else
+            {
+                Segments.Add(new ArraySegment<byte>(initialData));
+                Size = initialData.Length;
+                Capacity = initialData.Length;
+            }
+            _currentSegment = Segments[0];
+            _currentSegmentOffset = 0;
+        }
+
+        /// <summary>
+        /// Create a buffer from a list of byte arrays, it array is added as a segment to the buffer,
+        /// the Size and Capacity are set to the summ of the arrays lengths.</summary>
+        /// <param name="segments"></param>
+        public VectoredBuffer(byte[][] segments)
```

Comment:
I would prefer `arrays` here, and use segment(s) only to designate ArraySegment.